### PR TITLE
Show "Picture Books" in language card list (BL-12827)

### PR DESCRIPTION
**Pull request by andrew-polk**
This change is [Reviewable](https://reviewable.io/reviews/BloomBooks/BloomLibrary2/522)

### CODE REVIEW

**Comment by andrew-polk** [comment]
This has an issue; converted to draft

**Comment by andrew-polk** [comment]
Issue is resolved. Ready for review.

**Review by JohnThomson — CHANGES_REQUESTED** [review]
> Reviewed 5 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/BloomBooks/BloomLibrary2/522)* status: all files reviewed, 4 unresolved discussions (waiting on @andrew-polk)

___
*[`src/components/ByLanguageGroups.tsx` line 118 at r1](https://reviewable.io/reviews/BloomBooks/BloomLibrary2/522#-NiM4KRP5Bf3pSVGP753:-NiM4KRP5Bf3pSVGP754:b-t5jvjj) ([raw file](https://github.com/BloomBooks/BloomLibrary2/blob/43cf7a46427cbc1f299f2959157225fb5962a1e9/src/components/ByLanguageGroups.tsx#L118)):*
> ```XML
>             // eslint-disable-next-line no-loop-func
>             searchResults.books.forEach((book) => {
>                 let isFoundSomeLanguage = false;
> ```

For some reason I found this code quite hard to follow, maybe just because of jetlag. But I have a few ideas that might help.
Could we reverse the polarity here and let isMissingLanguage = true? I think the opposite polarity of these very similar flags (both end up telling us whether we found at least one of something) contributes to the difficulty of reading this code.
Actually I find it more difficult to reason about negative-named booleans, so I'd prefer hasSomeLananguage = false and hasLanguageForFeature = !needLangCheck, but it's a bit weird to claim that it automatically 'hasLanguageForFeature' when needLangCheck is false.
Another option: I think we could merge them into one variable, isBookInMap = false, and let addBookToLang set it true. Then, at the end, if the book didn't get added to any language, we add it to something, like you now do if isMissingLanguageForLanguageFeature. After all, if it has no language at all, that code should find no L1, and no book.languages[0]?.isoCode and will then add it to kTagForNoLanguage, just like you now do if isFoundSomeLanguage is false. So that might simplify the logic all around.
___
*[`src/components/ByLanguageGroups.tsx` line 188 at r1](https://reviewable.io/reviews/BloomBooks/BloomLibrary2/522#-NiM2zAeAdIKk6emLQYI:-NiM2zAeAdIKk6emLQYJ:byj56y1) ([raw file](https://github.com/BloomBooks/BloomLibrary2/blob/43cf7a46427cbc1f299f2959157225fb5962a1e9/src/components/ByLanguageGroups.tsx#L188)):*
> ```XML
>                             book.languages[0]?.isoCode ||
>                             kTagForNoLanguage
>                     );
> ```

So, the example in your comment isMissingLanguageForLanguageFeature, right, since the talking book feature is not present for Hako, the only language that the book thinks it has at all. Hako is presumably its lang1Tag, so we are going to add it to a list of books that are supposedly talking books in Hako.
Wouldn't it make more sense to find a way to add en and tpi arrays to the map so we could correctly show that it is talking in those languages?
Of course, another possibility is that the upload was configured to contain only Hako. Maybe the original book on the desktop was talking in en and tpi, but the user chose to only upload it in Hako, so in fact the uploaded version is not a talking book in any language? Perhaps we are just not filtering these tags when we upload a book with a subset of its languages?
If it's too dangerous, or more trouble than it's worth, to add the languages that the feature is set for to the map, I wonder if it wouldn't be better to just put it in kTagForNoLanguage? Then, for this book, we'd be saying that it is marked as a talking book, but not for any language we know about. (No, I guess if it has any languages we can't put it in that list because the filter below wouldn't find it.)
___
*[`src/components/ByLanguageGroups.tsx` line 191 at r1](https://reviewable.io/reviews/BloomBooks/BloomLibrary2/522#-NiM4w8IEdZds7WyMUQU:-NiM4w8IEdZds7WyMUQV:b-ruhtji) ([raw file](https://github.com/BloomBooks/BloomLibrary2/blob/43cf7a46427cbc1f299f2959157225fb5962a1e9/src/components/ByLanguageGroups.tsx#L191)):*
> ```XML
>                 }
>                 if (!isFoundSomeLanguage) {
>                     addBookToLang(kTagForNoLanguage);
> ```

Looks like it would tend to get in this list twice, since isMissingLanguageForLanguageFeature is always true if this is false and we needLangCheck. The check for 'similar' books in addBookToLang is probably preventing this, but I think it would be clearer and safer to use an else if here.
___
*[`src/model/Language.ts` line 71 at r1](https://reviewable.io/reviews/BloomBooks/BloomLibrary2/522#-NiMAsjYDrPQN5BqC_oD:-NiMAsjYDrPQN5BqC_oE:b-m1e57g) ([raw file](https://github.com/BloomBooks/BloomLibrary2/blob/43cf7a46427cbc1f299f2959157225fb5962a1e9/src/model/Language.ts#L71)):*
> ```XML
>             "book.detail.pictureBook",
>             "Picture Book (no text)"
>         );
> ```

Or it might be a sign language book that was uploaded without setting the language?

**Review by andrew-polk — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/BloomBooks/BloomLibrary2/522)* status: 4 of 5 files reviewed, 4 unresolved discussions (waiting on @JohnThomson)

___
*[`src/components/ByLanguageGroups.tsx` line 118 at r1](https://reviewable.io/reviews/BloomBooks/BloomLibrary2/522#-NiM4KRP5Bf3pSVGP753:-NiMa5zH7pJoVKoS66-p:b6zzz2o) ([raw file](https://github.com/BloomBooks/BloomLibrary2/blob/43cf7a46427cbc1f299f2959157225fb5962a1e9/src/components/ByLanguageGroups.tsx#L118)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

For some reason I found this code quite hard to follow, maybe just because of jetlag. But I have a few ideas that might help.
Could we reverse the polarity here and let isMissingLanguage = true? I think the opposite polarity of these very similar flags (both end up telling us whether we found at least one of something) contributes to the difficulty of reading this code.
Actually I find it more difficult to reason about negative-named booleans, so I'd prefer hasSomeLananguage = false and hasLanguageForFeature = !needLangCheck, but it's a bit weird to claim that it automatically 'hasLanguageForFeature' when needLangCheck is false.
Another option: I think we could merge them into one variable, isBookInMap = false, and let addBookToLang set it true. Then, at the end, if the book didn't get added to any language, we add it to something, like you now do if isMissingLanguageForLanguageFeature. After all, if it has no language at all, that code should find no L1, and no book.languages[0]?.isoCode and will then add it to kTagForNoLanguage, just like you now do if isFoundSomeLanguage is false. So that might simplify the logic all around.
</blockquote></details>

Thanks. I think that improves things considerably!
___
*[`src/components/ByLanguageGroups.tsx` line 188 at r1](https://reviewable.io/reviews/BloomBooks/BloomLibrary2/522#-NiM2zAeAdIKk6emLQYI:-NiMa8jTE6wkCCB6RhtV:b-mva9kl) ([raw file](https://github.com/BloomBooks/BloomLibrary2/blob/43cf7a46427cbc1f299f2959157225fb5962a1e9/src/components/ByLanguageGroups.tsx#L188)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

So, the example in your comment isMissingLanguageForLanguageFeature, right, since the talking book feature is not present for Hako, the only language that the book thinks it has at all. Hako is presumably its lang1Tag, so we are going to add it to a list of books that are supposedly talking books in Hako.
Wouldn't it make more sense to find a way to add en and tpi arrays to the map so we could correctly show that it is talking in those languages?
Of course, another possibility is that the upload was configured to contain only Hako. Maybe the original book on the desktop was talking in en and tpi, but the user chose to only upload it in Hako, so in fact the uploaded version is not a talking book in any language? Perhaps we are just not filtering these tags when we upload a book with a subset of its languages?
If it's too dangerous, or more trouble than it's worth, to add the languages that the feature is set for to the map, I wonder if it wouldn't be better to just put it in kTagForNoLanguage? Then, for this book, we'd be saying that it is marked as a talking book, but not for any language we know about. (No, I guess if it has any languages we can't put it in that list because the filter below wouldn't find it.)
</blockquote></details>

Yeah, it is hard to speculate exactly about what should be since we don't really know how things get into this state.
In the end, this is a case which represents bad data. And the main goal of the original code was to ensure the books *did* show on the page.
___
*[`src/components/ByLanguageGroups.tsx` line 191 at r1](https://reviewable.io/reviews/BloomBooks/BloomLibrary2/522#-NiM4w8IEdZds7WyMUQU:-NiMaOha1ctzvg-ADTdg:b-v7jw3o) ([raw file](https://github.com/BloomBooks/BloomLibrary2/blob/43cf7a46427cbc1f299f2959157225fb5962a1e9/src/components/ByLanguageGroups.tsx#L191)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Looks like it would tend to get in this list twice, since isMissingLanguageForLanguageFeature is always true if this is false and we needLangCheck. The check for 'similar' books in addBookToLang is probably preventing this, but I think it would be clearer and safer to use an else if here.
</blockquote></details>

No longer relevant with the change you suggested above.
___
*[`src/model/Language.ts` line 71 at r1](https://reviewable.io/reviews/BloomBooks/BloomLibrary2/522#-NiMAsjYDrPQN5BqC_oD:-NiMaWX62TiE8fVWus_c:bnfq0mc) ([raw file](https://github.com/BloomBooks/BloomLibrary2/blob/43cf7a46427cbc1f299f2959157225fb5962a1e9/src/model/Language.ts#L71)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Or it might be a sign language book that was uploaded without setting the language?
</blockquote></details>

I don't think that was actually possible with the previous code (though certain aberrant data conditions probably could allow it). But it certainly shouldn't be the case now.
Or at least, even if it is a book missing its feature tag, it is also a Picture Book.

**Review by JohnThomson — APPROVED** [review]
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/BloomBooks/BloomLibrary2/522)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @andrew-polk)
